### ISSUE TITLE
Random issue where unpack fails due to corrupt rar

### ISSUE THREAD

**Issue by Cpuroast** [sabnzbd/sabnzbd#2632]
Random issue where unpack fails due to corrupt rar but sab fails to catch it before it tries to unpack.

This has happened randomly on various nzbs, even on the built-in test nzbs.

I've manually run par2 checks on the archive and 1 or 2 archives are corrupted which par2 repairs, but sab never detected the error before it tried the unpack.

This is on 4.0.3 on macOS [[9fafe64](https://github.com/sabnzbd/sabnzbd/commit/9fafe64)], but it happened in previous 4.x versions.

Like the issue can happen, I remove the job, resubmit the same nzb and it unpacks correctly this time.

Does sab run a full par2 check after the download is complete before unpack?

I know that this is somewhat of a recent thing as I don't remember ever seeing this before.

I check the logs and quickcheck passes and everything before the unpack happens, then unpack fails and the webUI shows "Corrupt RAR file".

**Comment by Cpuroast** [comment]
Does sab run a full par2 check after the download is complete before unpack?

**Comment by Cpuroast** [comment]
I know that this is somewhat of a recent thing as I don't remember ever seeing this before.

**Comment by albino1** [comment]
I've been getting this occasionally as well.  I thought maybe it was something on my end, or possibly people uploading changed something in their setups.

**Comment by mnightingale** [comment]
> Does sab run a full par2 check after the download is complete before unpack?

The behaviour did change a bit in 4.0.0, the quickcheck to verify it switched to CRC32 instead of MD5 because it's less CPU intensive, I'm sure it's possible to calculate the chance of a hash collision of 32 vs 128 bit, but I'm sure it's staggeringly unlikely.

Based on your description I think SAB may benefit from a couple of changes:
- If unrar fails, repair and unrar again (make sure not to get into a unrar/repair loop)
- I'm not sure it already does this but if BadData is raised which results in increasing nzo.bad_articles then force a verify/repair

So as you know if the quick verify succeeds it won't do a full verify/repair.

However I have been meaning to make a PR which changes this part of code; because there is no guarantee a write call actually write everything we asked!

https://github.com/sabnzbd/sabnzbd/blob/9fafe64cff631ea6c61e65780f366046b2830108/sabnzbd/assembler.py#L183-L186

Basically, it should loop indefinitely until it's all written.

https://docs.python.org/3/library/io.html?highlight=write#io.RawIOBase.write

> Write the given [bytes-like object](https://docs.python.org/3/glossary.html#term-bytes-like-object), b, to the underlying raw stream, **and return the number of bytes written. This can be less than the length of b in bytes,** depending on specifics of the underlying raw stream, and especially if it is in non-blocking mode. (snip)

---

> The other question is, how are they getting corrupted in the 1st place, exact same nzb from the same servers, just on different runs.

Are you saying multiple runs of the same NZB sometimes succeeds?

If so, it does give more weight to their being less written than expected sometimes.

I'll make a PR.

**Comment by thezoggy** [comment]
Do you use direct unpack or not, try toggling whichever and retest with the nzb to see if outcome changes

**Comment by Cpuroast** [comment]
No direct unpack, haven't used it for a very long time.

**Comment by Cpuroast** [comment]
Maybe there should be a special option to do force a full verify/repair regardless of the result of the quickcheck.

**Comment by Cpuroast** [comment]
Switched to the new build which includes your new PR, will report back if I see it pop up again.

**Comment by mnightingale** [comment]
Develop does have a few substantial file writing changes, but understandably @Safihre wants to test them more extensively before a release.

Easy enough to backport to v4.0.x if required.

Very interested if it resolves your problem though 😄

**Comment by Cpuroast** [comment]
Doesn't look like it, I erased all of todays downloads and re-added them all to sab, was going well, until 
Corrupt RAR file" happened again.

**Comment by Cpuroast** [comment]
Checked the logs and once again:

2023-07-27 15:52:39,156::INFO::[newsunpack:1114] Quick-check for job is OK, skipping repair
2023-07-27 15:52:39,168::DEBUG::[postproc:808] Verified sets: {'job': True}

then proceeds to nuke the par2 files and unpack, and then fails.

**Comment by Cpuroast** [comment]
Sacrificing reliability in the name of speed isn't worth it, a full par2 verify/repair pass after each download is a good idea, like the good old days.

**Comment by Cpuroast** [comment]
Go figure, ran the same nzb 3-4 times and it now succeeded.

**Comment by sanderjo** [comment]
> Go figure, ran the same nzb 3-4 times and it now succeeded.

Intermittent? I've not read the whole thread, but could it be a hardware problem?

**Comment by mnightingale** [comment]
> Go figure, ran the same nzb 3-4 times and it now succeeded.

Which suggests to me the quick check is working, it'll calculate the same hashes evertime.

I'd concur with @sanderjo maybe a hardware issue?

Do you have another machine you could try or a different disk?

**Comment by Cpuroast** [comment]
I tried using the internal SSD and the external disk, the same issue randomly occurs on both.

It could be a hardware issue, but I don't experience any crashes or other weird behavior.

I know that when I grab a par2 file and run it against the broken set, it finds corrupted files.

**Comment by albino1** [comment]
Maybe try running something from an older 3.x branch against the same set of NZBs?

**Comment by Cpuroast** [comment]
Rolled back to 3.6.0 as that's the oldest in https://sabnzbd.org/download-stats, we'll see.

I know there's older on github, but I don't recall having this issues around a year ago.

**Comment by Cpuroast** [comment]
Well, not a guaranteed, but once again, erased and re-added all of today's downloads, this time using 3.6.0 [[b31fedd](https://github.com/sabnzbd/sabnzbd/commit/b31fedd)] and they all completed just fine.

**Comment by Cpuroast** [comment]
I just hammered 3.6.0 [[b31fedd](https://github.com/sabnzbd/sabnzbd/commit/b31fedd)] with 32x 1GB built-in test downloads and they all unpacked just fine, where as they would randomly fail to unpack on 4.0.3.

**Comment by mnightingale** [comment]
Did any on 3.6.0 need repair?

i.e. did any MD5 quick-check fail where CRC32 on newer version passes?
If they all succeeded the same as CRC32 does then I think we can rule out the download parts of the code and maybe investigate assembly, but I'm a bit stumped how quick-check could pass but still end up with a broken file.

It may be worth adding a special option to prefer the old MD5 quick(slow)-check at the expensive of CPU usage.

**Comment by Cpuroast** [comment]
No repairs needs, straight unpack, so far.

**Comment by Safihre** [comment]
@Cpuroast I don't know what to make of this.
I haven't see other reports of this on Reddit or our forums, so I am not sure where to look anymore.
One thing to try: run manual par2 on the failed downloads, does it find problems and in which file?

**Comment by Cpuroast** [comment]
Hi,

Yeah, it could be a combination of old hardware and more lenient CRC32 vs MD5.
In my testing since last week I've encountered it on 3.6.0 a few times, much less frequently than on recent versions.

And yeah, when I run manual par2 on the failed downloads by keeping all the par2s, it's able to identify the broken files and repair them, which sab doesn't try to do, because the quickcheck had already passed and it nuked the par2 files.

**Comment by Cpuroast** [comment]
Might be worth it to have an option to keep the par2 files around and if unpack fails because of "Corrupt RAR file", trigger a full par2 check/repair run before trying to unpack again.

**Comment by thezoggy** [comment]
> Might be worth it to have an option to keep the par2 files around until the very end, and if unpack fails because of "Corrupt RAR file", trigger a full par2 check/repair run before trying to unpack again.
> 
> And maybe as mentioned by @mnightingale have an option to use MD5 instead of CRC32.

the par files are kept around until the rar are no longer corrupt.. so does it fail to repair the files? 
which par are you using? par2cmdline? par2cmdline-turbo? or par2-sl64?

do you have debug logs of the whole event showing what par/rar showed?

**Comment by Safihre** [comment]
@thezoggy he already explained that we don't detect it's broken so it's deleted, that's indeed how it works.

@Cpuroast we already have the option to keep the source files by setting a download to +Unpack instead of +Delete.

**Comment by mnightingale** [comment]
It still sounds like a hardware issue to me, the chances of so many crc32/md5 collisions is practically impossible and the same thing downloaded multiple times won’t have different hashes meaning it works sometimes and not others.

Since it’s MacOS I would run [Apple diagnostics ](https://support.apple.com/en-gb/HT202731) then [MemTest86](https://www.memtest86.com/tech_creating-linux-mac.html) or MemTest86+ to test the memory (this will take hours)

**Comment by Safihre** [comment]
@ItsME6969 that doesn't sound like the same thing, as in the case of @Cpuroast SABnzbd doesn't detect it's broken. Here it does detect it's broken.
Do you have _Post process only verified jobs_ turned on in Config Switches?

**Comment by Safihre** [comment]
Indeed, try with that setting turned On if that catches things correctly.
Maybe also enable Debug logging so we can see why par2 wasn't triggered.

**Comment by Safihre** [comment]
@Cpuroast Any updated on this? Did you check your hardware?

**Comment by sanderjo** [comment]
So
```
Running external command: ['/usr/bin/par2', 'r' ...
libgomp: Thread creation failed: Resource temporarily unavailable
```

```
$ ldd /usr/bin/par2
        linux-vdso.so.1 (0x00007fffb37be000)
        libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x000079f85e037000)
        libgomp.so.1 => /lib/x86_64-linux-gnu/libgomp.so.1 (0x000079f85dff7000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x000079f85dfdd000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x000079f85dfbb000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x000079f85dde7000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x000079f85dca3000)
        /lib64/ld-linux-x86-64.so.2 (0x000079f85e28d000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x000079f85dc9b000)
```

Ah, there it is! OpenMP: https://en.wikipedia.org/wiki/OpenMP

So it this a resource restricted hardware? Or OS?

**Comment by thezoggy** [comment]
@ItsME6969 are you using par2cmdline-turbo? If so, what version.

**Comment by thezoggy** [comment]
> Sorry went afk for a second
> 
> @sanderjo that is all greek to me, mind dumbing it down a bit for me? :), And, yea, it 's a shared server but I've never had any problems resource-wise. I've only started noticing the issue recently.
> Specs
> ```
> -------------------- A Bench.sh Script By Teddysun -------------------
>  Version            : v2023-06-10
>  Usage              : wget -qO- bench.sh | bash
> ----------------------------------------------------------------------
>  CPU Model          : Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
>  CPU Cores          : 80 @ 2301.000 MHz
>  CPU Cache          : 51200 KB
>  AES-NI             : Enabled
>  VM-x/AMD-V         : Enabled
>  Total Disk         : 298.2 TB (139.0 TB Used)
>  Total Mem          : 346.3 GB (51.4 GB Used)
>  System uptime      : 80 days, 10 hour 0 min
>  Load average       : 7.76, 8.80, 8.88
>  OS                 : Ubuntu 22.04.2 LTS
>  Arch               : x86_64 (64 Bit)
>  Kernel             : 5.15.0-73-generic
>  TCP CC             : bbr
>  Virtualization     : Dedicated
>  IPv4/IPv6          : Online / Online
>  Organization       : AS60781 LeaseWeb Netherlands B.V.
>  Location           : Amsterdam / NL
>  Region             : North Holland
> ----------------------------------------------------------------------
> ```
> @thezoggy no I dont have that installed only;
> par2cmdline version 0.8.1, and 
> parpar --version 0.4.1
> 

par2cmdline uses openml and that release is 4y old.. there has been some commits since but no release. You might just need to reinstall par2cmdline to fixup it's depdencies.. as I assume your prob just got something funky going on with gcc/libomp.

personally  I'd say try removing par2cmdline and installing par2cmdline-turbo and trying again. It is significantly faster and takes advantage of modern hardware (parpar speedups). 

If your running sab in docker with Linux server, you can just load their docker mod to install it. If your managing things yourself you can grab release and install or compile yourself..

**Comment by Cpuroast** [comment]
things seems to have stabilized, could have been my external USB drive acting up or something with the previous version of Big Sur. I upgraded to 11.7.9 since and now I upgraded to 12.6.8 and things are behaving.

I progressively upgraded Sab from older to newer in the mean time, now at 3.7.2, seems ok. I'll stick with 3.7.2 for now until 4.1 final.

**Comment by Cpuroast** [comment]
As a side note I just tested 4.0.3 on 10.9 through 10.15 and it still runs fine.
Looks like Python.org will still require a minimum of 10.9 for Python 3.12, so unless something else drops support, Sab should still keep working.

**Comment by Cpuroast** [comment]
Quick update, while it did happen occasionally with 3.7.2, since upgrading back to 4.x this issue seems to happen more often. was recently on 4.1.0 Beta1 and it happened quite a few times, rolled back to 3.7.2 for now.

**Comment by Cpuroast** [comment]
It could be an interested test to make md5 like 3.x optional in 4.x.
Sacrificing integrity for speed is never a good thing.

**Comment by Safihre** [comment]
Based on all information here I'm 99% sure it doesn't have anything to do with MD5 vs CRC. 
The test downloads are not corrupt, they don't have missing articles or CRC errors in the articles. They should never need repair and they don't need repair with any other users.
_If_ there is a problem in SABnzbd it could be due to the writing being corrupt or something along those lines.
But the most likely culprit is your hardware, especially because you experienced it only all SABnzbd version and you mentioned it's a USB storage.

**Comment by Cpuroast** [comment]
I switched to local SSD storage for Temp and USB for Complete, although during my initial testing, I was also seeing issues with "corrupt" rars when unpacking from the SSD, so I still don't know.
Back on 3.7.2 and so far so good.
